### ISSUE TITLE
FAC17 mentor feedback: Week 4

### ISSUE THREAD

**Issue by dubhcait** [foundersandcoders/master-reference#1000]
#### Server basics

- Understand what is a server
- Understand the relationship between a server and a client

#### Node server basics

- Understand what is Node.js
- Understand how running Node differs from running JavaScript in your browser
- Understand basic ES6 syntax and general file system manipulation
- Understand what a node module is and the different types of modules
- Understand and use basic core modules (e.g. `http`, `fs`, `querystring`, `url`) and some of their properties and methods
- Understand the structure of a URL
- Understand the `request` and `response` objects and some of their properties and methods
- Understand how to set up a server with different endpoints and http methods (e.g. `GET`, `POST`)

#### Module design pattern and server file structure

- Understand what is npm
- Understand what is a package.json and semantic versioning
- Understand how to import functions and objects from other modules using `require`.
- Understand how to export modules using `module.exports`
- Be introduced to writing code into reusable pieces
- Understand what is linting and why it is important
- Understand a basic structure for routing logic and modularisation
- Get familiar with a typical file structure on the server side (e.g. which files may go into the `public` and `src` folders)

#### Testing

- Understand how to use tests to make non-breaking changes to an existing codebase

### Revision questions

- What is a server and what is Node.js?
- What types of Node modules are there?
- How do you set up a Node server that listens to different endpoints?
- What is a typical server file structure?
- How do you import (`require`) and export modules (`module.exports`)?
- How are you testing your server functionality using `tape`?

--- 

### More Emphasis
- open the file with live preview, then each stage show the browser ( pre and post public route) 
- What the packages are (refer to tape, nodemon, tap-spec etc)
- module.exports => require (the link of the two)
- demo REPL, npm search 
- Draw everything together afterwards on a white board - individual parts
  - emphasis the requests made by the html file (img, css, js)
  
---
  
 ## Node girls workshop

 
 - review node girl express workshop and notes => combine 
 -  tell them in step 10 - remember week 3
 
 ![](https://i.ibb.co/5Ld0NXV/Screenshot-2019-05-20-at-11-53-40.png)

---

## ES6 morning challenge
- after workshop - have link to ES6 cheat sheet

---

## Modularisation morning challenge
- seems fine

---

## Research 

- Topic 1: Architecting
- Topic 2: Engineering (Asyncronous functions)
- Topic 3: Packaging
- Topic 4: Deploying ( make sure they can and share their example) Or ? workshop moved to this week? https://hackmd.io/SM9XYfssSIWjgwC20yDslQ?view

**Comment by Baalwaan** [comment]
## Wk 4 - Node (Wk1) Cohort Feedback

- Anna 
    - Introduction to topic a MUST.

- Banna 
    - What is node? an introduction would be more beneficial and help cohort get some understanding of what it is.
    - Visualisation important and helpful !!!

- Bobby 
    - Heroku workshop maybe before research (or maybe scrap the whole idea first and introduce deploying following week)
    - Bobby - Where console.log appear in browser or terminal ? 

- Henry 
    - stdout.write ?

- James 
    - Research topics not material to use for revision, because topics may not be written to the highest standards

**Comment by charlielafosse** [comment]
Hi, thanks for your feedback! Check out the [curriculum update process](https://github.com/foundersandcoders/master-reference/blob/master/curriculum-planning/curriculum-update-process.md) if you need a refresher on how we review and update the curriculum.

If changes are **tiny** I'll suggest that you raise a **PR** - assign me and I can help out. For **minor** or **major** changes you should raise an **issue** if you feel really strongly about them, and assign me. Make sure to have a search first and check that you're not duplicating an issue or PR that already exists. 

Be aware that major changes may well not happen before you deliver your week - so raise issues sooner rather than later! If an issue or PR relates to a specific workshop or challenge, it should be raised within the same repo as the related content (this might not always be `master-reference`).

---
> Learning outcomes don't reflect the workshops of the week

Could you expand on this a little bit?

>after workshop - have link to ES6 cheat sheet

Cool idea. If you make one or find one, one it'd be a great thing to add to Week 4's [resources folder](https://github.com/foundersandcoders/master-reference/tree/master/coursebook/week-4/resources) - make a PR and assign me if so!

>Anna
Introduction to topic a MUST.
Banna
What is node? an introduction would be more beneficial and help cohort get some understanding of what it is.
Visualisation important and helpful !!!

A need for more of an intro/oversight to topics and projects has been a common theme with FAC16 - and I think that it's a really good idea. This wouldn't necessarily be a change to the curriculum, but rather something you might want to prepare and deliver yourself. 
However, if you do create additional introductory content that you think should be merged into the curriculum, then this would likely be a **minor change**. Feel free to raise an issue, and assign me.

> Heroku workshop maybe before research (or maybe scrap the whole idea first and introduce deploying following week)

I agree that there should definitely be space in the curriculum for understanding Heroku a bit better before we start trying to implement it. I know that @Oliversw and @teenie-quaggard did something in [Week 5](https://github.com/foundersandcoders/master-reference/issues/936) - see presentation notes [here](https://hackmd.io/SM9XYfssSIWjgwC20yDslQ?view) - but it'd be even better to transplant this over to Week 4. Since there's already [an issue](https://github.com/foundersandcoders/master-reference/issues/977) out there, contribute to that and if something's worth merging, open a PR!

---

Let me know if you've got any questions, or if I've missed anything out. Enjoy the mentoring! You'll be great :sun_with_face:
